src/Symfony/Component/VarDumper/Dump2.php

```diff
@@ -0,0 +1,270 @@
+ */
+
+namespace Symfony\Component\VarDumper;
+
+use Symfony\Component\VarDumper\Cloner\VarCloner;
+use Symfony\Component\VarDumper\Dumper\CliDumper;
+use Symfony\Component\VarDumper\Dumper\HtmlDumper;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class Dump2
+{
+    const AS_HTML = 'html';
+    const AS_CLI = 'cli';
+    const AS_TEXT = 'text';
+
+    private $vars;
+    private $dumpOnDestruct;
+    private $to = null;
+    private $format = 'env';
+    private $die = false;
+    private $title = '';
+    private $trace = false;
+    private $maxItems = 2500;
+    private $maxDepth = -1;
+    private $maxItemsPerDepth = -1;
+    private $maxStringLength = -1;
+    private $maxStringWidth = -1;
+    private $useRefHandles = true;
+    private $casters = array();
+    private $replaceCasters = false;
+    private $filter = 0;
+
+    public function __construct(array $vars, $dumpOnDestruct = true)
+    {
+        $this->vars = $vars;
+        $this->dumpOnDestruct = $dumpOnDestruct;
+    }
+
+    public function __destruct()
+    {
+        if ($this->dumpOnDestruct) {
+            $this->dumpOnDestruct = false;
+            $this->dump();
+        }
+    }
+
+    public function dump()
+    {
+        if ($this->replaceCasters) {
+            $cloner = new VarCloner($this->casters);
+        } else {
+            $cloner = new VarCloner();
+            $cloner->addCasters($this->casters);
+        }
+        $cloner->setMaxItems($this->maxItems);
+        $cloner->setMaxString($this->maxStringLength);
+
+        $dumper = $this->getDumper($this->getFormat($this->format), $this->output);
+        $dumper->setMaxStringWidth($this->maxStringWidth);
+
+        foreach ($this->vars as $v) {
+            $dumper->dump($cloner->cloneVar($v, $this->filters)
+                ->withMaxDepth($this->maxDepth)
+                ->withMaxItemsPerDepth($this->maxItemsPerDepth)
+                ->withRefHandles($this->useRefHandles)
+            );
+        }
+
+        if ($this->die) {
+            exit(1);
+        }
+    }
+
+    protected function getFormat($format)
+    {
+        foreach (array($format, getenv('DUMP_AS')) as $format) {
+            $format = strtolower($format);
+            if ('html' === $format || 'cli' === $format || 'text' === $format) {
+                return $format;
+            }
+        }
+
+        return 'cli' === PHP_SAPI ? 'cli' : 'html';
+    }
+
+    protected function getDumper($format, $output)
+    {
+        if ('html' === $format) {
+            return new HtmlDumper($output);
+        }
+
+        $dumper = new CliDumper($output);
+
+        if ('text' === $format) {
+            $dumper->setColors(false);
+        }
+
+        return $dumper;
+    }
+
+    /**
+     * Sets the output destination of the dump.
+     *
+     * @param callable|resource|string $output A line callable, an opened stream or an output path.
+     */
+    public function to($output)
+    {
+        $dump = clone $this;
```

Comment:
@Taluu which doesn't mean it's not a value object :) Do you have any argument that would balance towards a mutable interface? I started with an immutable one because it looked fine to me but of course this can be changed. If it's only a matter of preference, then the author (me) wins ;)